torch/_inductor/kernel/flex_decoding.py

```diff
@@ -90,12 +95,39 @@ def flex_decoding_grid(batch_size, num_heads, n_keys, d_model, meta):
     TILE_KV_OG = tl.cdiv(KV_LEN, SPLIT_KV)
     TILE_KV = tl.cdiv(TILE_KV_OG, BLOCK_N) * BLOCK_N
 
+    TILE_KV_MULTIPLE: tl.constexpr = (TILE_KV // BLOCK_N)
+
+    SPARSE_Z = {{size("KV_NUM_BLKS", 0)}}
+    SPARSE_H = {{size("KV_NUM_BLKS", 1)}}
+
+    tl.static_assert(SPARSE_KV_BLOCK_SIZE >= BLOCK_N and SPARSE_KV_BLOCK_SIZE % BLOCK_N == 0)
+    SPARSE_KV_MULTIPLE: tl.constexpr = (SPARSE_KV_BLOCK_SIZE // BLOCK_N)
+    SPARSE_KV_BLOCK_CNT: tl.constexpr = KV_LEN // SPARSE_KV_BLOCK_SIZE
+
     MATMUL_PRECISION = Q.dtype.element_ty
 
     off_z = tl.program_id(0) // H
     off_h = tl.program_id(0) % H
     off_t = tl.program_id(1)
-    off_n = off_t * TILE_KV
+
+    sparse_idx_z = off_z % SPARSE_Z
+    sparse_idx_h = off_h % SPARSE_H
+
+    sparse_hz_offset = sparse_idx_z * SPARSE_H + sparse_idx_h
```

Comment:
I feel like we should  be using strides instead of the size of the dim of the, that beings said I think its okay to do this, as long as we assert they are contiguous(they should be). I know Horace added some code around this in the refactor PR

Reply:
It's not in the refactor yet, so I'll skip for this PR. With that said, I think it's neccessary  for things like, e.g. packing multiple layers of precomputed masks.